libtiledbvcf/src/read/reader.cc

```diff
@@ -894,6 +1186,258 @@ std::vector<SampleAndId> Reader::prepare_sample_names() const {
   return result;
 }
 
+std::vector<SampleAndId> Reader::prepare_sample_names_v4() const {
+  std::vector<SampleAndId> result;
+
+  for (const std::string& s : params_.sample_names) {
+    std::string name;
+    if (!VCFUtils::normalize_sample_name(s, &name))
+      throw std::runtime_error(
+          "Error preparing sample list for export; sample name '" + s +
+          "' is invalid.");
+
+    const auto& sample_ids = dataset_->metadata().sample_ids;
+    auto it = sample_ids.find(name);
+    if (it == sample_ids.end())
+      throw std::runtime_error(
+          "Error preparing sample list for export; sample '" + s +
+          "' has not been registered.");
+
+    result.push_back({.sample_name = name, .sample_id = it->second});
+  }
+
+  if (!params_.samples_file_uri.empty()) {
+    const auto& metadata = dataset_->metadata();
+    auto per_line = [&metadata, &result](std::string* line) {
+      std::string name;
+      if (!VCFUtils::normalize_sample_name(*line, &name))
+        throw std::runtime_error(
+            "Error preparing sample list for export; sample name '" + *line +
+            "' is invalid.");
+
+      const auto& sample_ids = metadata.sample_ids;
+      auto it = sample_ids.find(name);
+      if (it == sample_ids.end())
+        throw std::runtime_error(
+            "Error preparing sample list for export; sample '" + *line +
+            "' has not been registered.");
+
+      result.push_back({.sample_name = name, .sample_id = it->second});
+    };
+    utils::read_file_lines(*vfs_, params_.samples_file_uri, per_line);
+  }
+
+  // No specified samples means all samples.
+  if (result.empty()) {
+    const auto& md = dataset_->metadata();
+    for (const auto& s : md.sample_names) {
+      auto it = md.sample_ids.find(s);
+      if (it == md.sample_ids.end())
+        throw std::runtime_error(
+            "Error preparing sample list for export; sample '" + s +
+            "' has not been registered.");
+      result.push_back({.sample_name = s, .sample_id = it->second});
+    }
+  }
+
+  return result;
+}
+
+void Reader::prepare_regions_v4(
+    std::vector<Region>* regions,
+    std::unordered_map<std::string, std::vector<size_t>>*
+        regions_index_per_contig,
+    std::vector<std::pair<std::string, std::vector<QueryRegion>>>*
+        query_regions) const {
+  const uint32_t g = dataset_->metadata().anchor_gap;
+  // Use a linked list for pre-partition regions to allow for parallel parsing
+  // of BED file
+  std::list<Region> pre_partition_regions_list;
```

Comment:
I'm not sure what a `list` buys us over a `vector` here. Note that if this contains a large number of elements, the destruction time can be significantly larger because it needs to deallocate each individual element rather than deallocating a single virtually-contiguous buffer in a vector.

Reply:
The reason we use a list is in the `Region::parse_bed_file_htslib` we parallelize the parsing on a per contig bases. We then `splice` the lists together without copying. We eventually sort/filter the regions below and produce a vector which we carry through the entire read.